src/Tools/BuildValidator/Program.cs

```diff
@@ -134,6 +152,57 @@ static int HandleCommand(string[] assembliesPath, string sourcePath, string[]? r
             }
         }
 
+        private static AssemblyInfo[] GetAssemblyInfos(
+            IEnumerable<string> assemblySearchPaths,
+            IEnumerable<string> excludes,
+            ILogger logger)
+        {
+            var map = new Dictionary<Guid, AssemblyInfo>();
+            foreach (var directory in assemblySearchPaths)
+            {
+                foreach (var filePath in getAssemblyPaths(directory))
+                {
+                    if (excludes.Any(x => filePath.IndexOf(x, FileNameEqualityComparer.StringComparison) >= 0))
```

Comment:
>excludes.Any(x => filePath.IndexOf(x, FileNameEqualityComparer.StringComparison) >= 0) [](start = 24, length = 86)

Could this be expensive if there are many excludes and many assemblies in the search paths? Perhaps create a `HashSet<string>` for the excludes in this method.

Reply:
How would a `HashSet<string>` help here? In this case it's a full file path that we have to check for a substring match. Not sure how we can cache that. The returned file paths should already be unique given the way the file system is walked here.

Reply:
Ah, I missed the fact we're matching substrings here.

---
In reply to: [589655092](https://github.com/dotnet/roslyn/pull/51694#discussion_r589655092) [](ancestors = 589655092)